src/shared/api.ts

```diff
@@ -0,0 +1,265 @@
+import { queryGraphQL } from '../graphql'
+import { createAggregateError } from './errors'
+
+const discussionCommentFieldsFragment = `
+    fragment DiscussionCommentFields on DiscussionComment {
+      id
+      author {
+        ...UserFields
+      }
+      contents
+      html
+      inlineURL
+      createdAt
+      updatedAt
+    }
+  `
+
+function discussionThreadFieldsFragment(relativeRev?: string): string {
+    let relativeRevFields = ''
+    if (relativeRev) {
+        relativeRevFields = `
+        relativePath(rev: $relativeRev)
+        relativeSelection(rev: $relativeRev) {
+          startLine
+          startCharacter
+          endLine
+          endCharacter
+        }
+    `
+    }
+    return `
+    fragment DiscussionThreadFields on DiscussionThread {
+      id
+      author {
+        ...UserFields
+      }
+      title
+      target {
+        __typename
+        ... on DiscussionThreadTargetRepo {
+          repository {
+            name
+          }
+          path
+          branch {
+            displayName
+          }
+          revision {
+            displayName
+          }
+          selection {
+            startLine
+            startCharacter
+            endLine
+            endCharacter
+            linesBefore
+            lines
+            linesAfter
+          }
+          ${relativeRevFields}
+        }
+      }
+      inlineURL
+      createdAt
+      updatedAt
+      archivedAt
+    }
+
+    fragment UserFields on User {
+      displayName
+      username
+      avatarURL
+    }
+  `
+}
+
+/**
+ * Creates a new discussion thread.
+ *
+ * @return Promise that emits the new discussion thread.
+ */
+export async function createThread(
+    input: SourcegraphGQL.IDiscussionThreadCreateInput,
+    relativeRev = ''
+): Promise<SourcegraphGQL.IDiscussionThread> {
+    const { data, errors } = await queryGraphQL(
+        `
+        mutation CreateThread($input: DiscussionThreadCreateInput!, $relativeRev: String!) {
+          discussions {
+            createThread(input: $input) {
+              ...DiscussionThreadFields
+              comments {
+                totalCount
+                nodes {
+                  ...DiscussionCommentFields
+                }
+              }
+            }
+          }
+        }
+        ${discussionThreadFieldsFragment(relativeRev)}
+        ${discussionCommentFieldsFragment}
+      `,
+        { input, relativeRev }
+    )
+    if (!data || !data.discussions || !data.discussions.createThread) {
+        throw createAggregateError(errors)
```

Comment:
We have nicer error handling now using `dataOrThrowErrors`! See https://sourcegraph.sgdev.org/github.com/sourcegraph/sourcegraph/-/blob/src/backend/graphql.tsx#L21:17 and examples in https://sourcegraph.sgdev.org/github.com/sourcegraph/sourcegraph/-/commit/78e0791ca3e3d4ac37d13b38c6da84473cb6c4e5